Add static backdrop to offcanvas

chefarbeiter opened the pull request:
Adds static backdrop to the offcanvas component. In analogy to the modal, the `hidePrevented` event is triggered for prevented closing actions. The event is also triggered when the user presses the escape key and `keyboard` is set to `false`.

Fixes #34247


- [x]  consider move the doc section

chefarbeiter committed:
Add static backdrop to offcanvas

chefarbeiter committed:
Trigger prevented event on esc with keyboard=false

GeoSot commented:
@twbs/css-review  @twbs/js-review do we want this feature? May I continue reviewing it?

For the record, I would prefer a way, where the config value to be of one type only (not string OR bool)

mdo commented:
I'm into it—we should provide the same functionality as the modal component.

GeoSot commented:
Ok , lets go. 
@chefarbeiter do you believe we need the `hidePrevented`? Can we avoid it for now?

Secondly I know you have replicated the tests of modal, but please try to cover all of your code https://coveralls.io/builds/46517959

GeoSot commented on js/tests/unit/offcanvas.spec.js:
```diff
@@ -90,19 +105,70 @@ describe('Offcanvas', () => {
 
-      document.dispatchEvent(keyDownEsc)
+        const expectEnd = () => {
+          setTimeout(() => {
```
it would be better if we could omit `setTimeout` usage

chefarbeiter replied:
I removed the `setTimeout` usages

chefarbeiter commented:
> do you believe we need the `hidePrevented`? Can we avoid it for now?

@GeoSot I have also taken the event from the modal component. There it was introduced with #29516. In the PR an animation was added when the user clicks in the static backdrop. I didn't take over this feature for the offcanvas. So I think the right answer to your question is: If we leave it that there is no animation in the offcanvas, we don't need the event. If we take over the animation from the modal component, we should also keep the event.

chefarbeiter committed:
Fix unit test: apply event to the right element

chefarbeiter commented:
> Secondly I know you have replicated the tests of modal, but please try to cover all of your code https://coveralls.io/builds/46517959

Fixed. There was even an existing test (not by me) that was supposed to cover exactly this case, but due to a small bug it didn't.

chefarbeiter committed:
Remove setTimeout call from new tests

GeoSot commented:
@patrickhlauke Can you check for any grammar fixes please? 

And tell us your opinion, if the option example needs to be along with the other [three backdrop examples ](https://deploy-preview-35832--twbs-bootstrap.netlify.app/docs/5.1/components/offcanvas/#backdrop)

patrickhlauke commented on site/content/docs/5.1/components/offcanvas.md:
```diff
@@ -79,6 +79,28 @@ You can use a link with the `href` attribute, or a button with the `data-bs-targ
 
+### Static backdrop
+
+When backdrop is set to static, the offcanvas will not close when clicking outside it.
```
`outside *of* it`

patrickhlauke commented on site/content/docs/5.1/components/offcanvas.md:
```diff
@@ -79,6 +79,28 @@ You can use a link with the `href` attribute, or a button with the `data-bs-targ
+  </div>
+  <div class="offcanvas-body">
+    <div>
+      I will not close if you click outside me.
```
`outside *of* me`

patrickhlauke commented on site/content/docs/5.1/components/offcanvas.md:
```diff
@@ -225,7 +247,7 @@ Options can be passed via data attributes or JavaScript. For data attributes, ap
 | Name | Type | Default | Description |
 | --- | --- | --- | --- |
-| `backdrop` | boolean | `true` | Apply a backdrop on body while offcanvas is open |
+| `backdrop` | boolean or the string `static` | `true` | Apply a backdrop on body while offcanvas is open. Alternatively, specify `static` for a backdrop which doesn't close the offcanvas on click. |
```
`on click` > `when clicked`

patrickhlauke commented on site/content/docs/5.1/components/offcanvas.md:
```diff
@@ -266,6 +288,7 @@ Bootstrap's offcanvas class exposes a few events for hooking into offcanvas func
 | `shown.bs.offcanvas` | This event is fired when an offcanvas element has been made visible to the user (will wait for CSS transitions to complete). |
 | `hide.bs.offcanvas` | This event is fired immediately when the `hide` method has been called. |
 | `hidden.bs.offcanvas` | This event is fired when an offcanvas element has been hidden from the user (will wait for CSS transitions to complete). |
+| `hidePrevented.bs.offcanvas` | This event is fired when the offcanvas is shown, its backdrop is `static` and a click outside the offcanvas or an escape key press  is performed with the keyboard option or `data-bs-keyboard` set to `false`. |
```
```
... `static`, and a click outside of the ...
```

there''s a stray extra space after `press`

that last bit of sentence was very confusing though - the various and and multiple or clauses make it quite unclear what is and isn't mutually exclusive here. worth revisiting/clarifying a bit more? do those or clauses apply just to the "a click outside of the offcanvas or an exscape key press is performed" ?

chefarbeiter committed:
Fix offcanvas doc

GeoSot commented:
@patrickhlauke  On modal case, when we click on a static backdrop, we can see the modal making a scaling reaction.

Should we do something similar here to help user understand?

patrickhlauke commented:
> Should we do something similar here to help user understand?

with the modal, of sort of works as it's "floating" freely on the page. for the offcanvas side panel, i think it would look a bit odd if it somehow "stretched/bounced" out towards the overlay area. it's a more solid component anchored onto the side of the browser window, rather than being floaty...

(sorry, lots of words to say "nah")

GeoSot commented:
> And tell us your opinion, if the option example is better to be along with the other [three backdrop examples], or not (https://deploy-preview-35832--twbs-bootstrap.netlify.app/docs/5.1/components/offcanvas/#backdrop)

Last thing to go

mdo commented:
@GeoSot @chefarbeiter Maybe we split all those into their own examples with clear labels? Feels a little crowded and not as easy to scan three examples in one :).

chefarbeiter committed:
Move backdrop examples to examples section

chefarbeiter commented:
I implemented @mdo's suggestion and converted the backdrop examples into standalone examples. Since one of the examples was the default behavior that is also in all the other examples on the page, I did not adopt it but removed it.

GeoSot commented:
Thank you a lot @chefarbeiter. Nice job 👏 

 _Please, if it is not a great effort can you 'fix' the modal test to on another MR?_

chefarbeiter commented:
> _Please, if it is not a great effort can you 'fix' the modal test on another MR?_

@GeoSot I have checked it. The test for the modal component does not contain this error. There the event is triggered correctly in the dialog and not in the document:
https://github.com/twbs/bootstrap/blob/b59b1080e1324394331b670fb102343ecb8736f8/js/tests/unit/modal.spec.js#L423-L426

folknor commented:
@GeoSot Is anything from this pull request missing? It looks to me like it can be closed.
